onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -54,6 +55,46 @@ static common::Status AllocateBufferUsingDeviceAllocatorFromShapeAndType(const T
   return Status::OK();
 }
 
+// deleter for external data tensors managed by an OrtValue; manages the release of
+// the tensor's data buffer (which points to the external data) and the tensor itself
+struct ExtDataValueDeleter {
+  OrtCallback ext_delete_cb;
+  Tensor* p_tensor;
+  void operator()(void*) noexcept {
+    ext_delete_cb.f(ext_delete_cb.param);
+    delete p_tensor;
+  }
+};
+
+// given a tensor proto with externdal data return an OrtValue with a tensor for
+// that data; the pointers for the tensor data and the tensor itself are owned
+// by the OrtValue's deleter
+static inline common::Status ExtDataTensorProtoToTensor(const Env& env,
+                                                        const std::basic_string<PATH_CHAR_TYPE>& proto_path,
+                                                        const ONNX_NAMESPACE::TensorProto& tensor_proto,
+                                                        Tensor& tensor, OrtCallback& ext_data_deleter) {
+  ORT_ENFORCE(utils::HasExternalData(tensor_proto));
+  ORT_ENFORCE(!proto_path.empty());
+
+  void* ext_data_buf = nullptr;
+  SafeInt<size_t> ext_data_len = 0;
+  ORT_RETURN_IF_ERROR(utils::GetExtDataFromTensorProto(env, proto_path.c_str(), tensor_proto,
+                                                       ext_data_buf, ext_data_len, ext_data_deleter));
+
+  // NB: creating a do-nothing allocator per tensor is wasteful; can perhaps be
+  // avoided if the Tensor class implements the do-nothing behavior when given a
+  // nullptr for the allocator argument
+  const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto.data_type())->GetElementType();
+  std::vector<int64_t> tensor_shape_vec = utils::GetTensorShapeFromTensorProto(tensor_proto);
+  TensorShape tensor_shape{tensor_shape_vec};
+
+  auto p_tensor = std::make_unique<Tensor>(type, tensor_shape, ext_data_buf, OrtMemoryInfo(CPU,
+                                           OrtAllocatorType::OrtDeviceAllocator));
+  tensor = std::move(*p_tensor);
```

Comment:
Should be able to update the Tensor more directly instead of going via a unique_ptr.

`tensor = std::move(Tensor(type, tensor_shape, ext_data_buf, OrtMemoryInfo(CPU, OrtAllocatorType::OrtDeviceAllocator)));`

Reply:
When I try `tensor = std::move(Tensor(...))` clang gives this warning`Moving a temporary object prevents copy elision`.
I figure we can do this instead `tensor = Tensor(...)` with no loss in performance.

Reply:
Cool. I suspected the std::move wasn't necessary but couldn't remember for sure.